edge-modules/api-proxy-module/src/main.rs

```diff
@@ -0,0 +1,125 @@
+	let notify_received_config = notify_need_reload_api_proxy.clone();
+	let notify_certs_rotated = notify_need_reload_api_proxy.clone();
+
+	let shutdown_signal_twin_state = Arc::new(Notify::new());	
+	let shutdown_signal_config_monitor = Arc::new(Notify::new());
+	let shutdown_signal_cert_monitor = Arc::new(Notify::new());
+	let shutdown_signal_loop_stask = Arc::new(Notify::new());
+
+	let client = monitors::config_monitor::get_sdk_client();
+	let mut shutdown_handle = client.inner().shutdown_handle().expect("couldn't get shutdown handle");
+
+
+	let report_twin_state_handle = client.report_twin_state_handle();
+
+	let twin_state_poll_task = monitors::config_monitor::poll_twin_state(report_twin_state_handle, shutdown_signal_twin_state.clone());
+	let config_monitor_task = monitors::config_monitor::start(client , notify_received_config, shutdown_signal_config_monitor.clone());
+	let cert_monitor_task = monitors::certs_monitor::start(notify_certs_rotated, shutdown_signal_cert_monitor.clone());
+	let loop_task = nginx_controller_loop(notify_need_reload_api_proxy, shutdown_signal_loop_stask.clone());
+	
+
+	//Switch atomic var to true to tell all task to shutdown.
+	ctrlc::set_handler(move ||  {
+		block_on(shutdown_handle.shutdown()).expect("Could not shutdown gracefully");
+		shutdown_signal_twin_state.notify();
+		shutdown_signal_config_monitor.notify();
+		shutdown_signal_cert_monitor.notify();
+		shutdown_signal_loop_stask.notify();
+    }).expect("Error setting Ctrl-C handler");
+
+
+	//kill all client first
+	futures::future::join_all(vec![
+		Box::pin(twin_state_poll_task) as Pin<Box<dyn Future<Output = ()>>>,
+		Box::pin(config_monitor_task) as Pin<Box<dyn Future<Output = ()>>>,
+		Box::pin(cert_monitor_task) as Pin<Box<dyn Future<Output = ()>>>,
+		Box::pin(loop_task) as Pin<Box<dyn Future<Output = ()>>>,
+	]).await;
+
+	log::warn!("Gracefully exiting");
+}
+
+pub async fn nginx_controller_loop(notify_need_reload_api_proxy: Arc<Notify>, shutdown_signal: Arc<Notify>){
+	let program_path= "/usr/sbin/nginx";
+	let args = vec!["-c".to_string(), "/app/nginx_config.conf".to_string(),"-g".to_string(),"daemon off;".to_string()];
+	let name = "nginx";
+	let stop_proxy_name = "stop nginx";
+	let stop_proxy_program_path = "nginx";
+	let stop_proxy_args = vec!["-s".to_string(), "stop".to_string()];
+
+
+	//Wait for certificate rotation and for parse configuration.
+	//This is just to avoid error at the beginning when nginx tries to start
+	//but configuration is not ready
+	notify_need_reload_api_proxy.notified().await;
+	notify_need_reload_api_proxy.notified().await;
+
+	loop{
+		//Make sure proxy is stopped by sending stop command. Otherwise port will be blocked
+		let command = Command::new(stop_proxy_program_path).args(&stop_proxy_args)
+		.spawn()
+		.with_context(|| format!("Failed to start {:?} process.", stop_proxy_name)).expect("Cannot stop proxy!");
+		command.await.expect("Error while trying to wait on stop proxy future");
```

Comment:
This is required, if we can't stop Nginx, we cannot start a new nginx process because it opens a port.
We need to be able to stop it if we want to start a new nginx process.
If we can't stop the current Nginx process, we are basically frozen.